### PR TITLE
Implement 'trim()' scalar string function

### PR DESCRIPTION
**Pull request by smakalias**
## Summary of the changes / Why this improves CrateDB

Implement 'trim()' scalar string function

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

File: sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            DataTypes.STRING);
+    }
+
+    @Nullable
+    private static String trim(String mode, String charsToTrim, String value) {
+        if (value == null || charsToTrim == null) {
+            return value;
+        }
+
+        final HashSet<Character> trimmableChars = new HashSet<>(Chars.asList(charsToTrim.toCharArray()));
```
Comment (smakalias) [inline sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java]: @mfussenegger I'm thinking of adding an optimized version for single char strings as I think this use case is heavily used. Do you think it's worth it?

File: sql-parser/src/main/antlr/SqlBase.g4 [hunk 2]
```diff
@@ -693,6 +695,10 @@ ESCAPE: 'ESCAPE';
 ASC: 'ASC';
 DESC: 'DESC';
 SUBSTRING: 'SUBSTRING';
+TRIM: 'TRIM';
```
Comment (mfussenegger) [inline sql-parser/src/main/antlr/SqlBase.g4]: Can we mark these as non-reserved? Otherwise this is a breaking change.

File: sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java [hunk 3]
```diff
@@ -347,6 +347,45 @@ public void testUserFunctions() {
         assertInstanceOf("USER", FunctionCall.class);
     }
 
+    @Test
+    public void testTrimFunctionAllArgs() {
+        assertInstanceOf("TRIM(LEADING 'A' FROM chars)", FunctionCall.class);
+        assertInstanceOf("TRIM(TRAILING 'A' FROM chars)", FunctionCall.class);
```
Comment (mfussenegger) [inline sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java]: Wouldn't it make sense to also assert how the `FunctionCall` looks like?

File: sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java (hunk 1 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java]: I guess it would be worth it. That would mean to initiate a `FunctionCall` with 2 args in the `AstBuilder` so that we can already dispatch to the optimized version instead of having to do a check on `charsToTrim` on runtime for each row, right?

File: sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Nullable
+    private static String trim(@Nonnull String mode,
+                               @Nullable String charsToTrim,
+                               @Nullable String value) {
+        if (value == null || charsToTrim == null) {
+            return value;
+        }
+
+        final HashSet<Character> trimmableChars = new HashSet<>(Chars.asList(charsToTrim.toCharArray()));
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java]:
```suggestion
        final HashSet<Character> trimmableChars = Set.of<>(charsToTrim.toCharArray());
```

File: sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            while (start < len && trimmableChars.contains(value.charAt(len - 1))) {
+                len--;
+            }
+        }
+
+        return value.substring(start, len);
+    }
+
+    private static boolean isLeadingOrBothMode(@Nonnull String mode) {
+        return mode.equals("LEADING") || mode.equals("BOTH");
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java]: Could we make this more strict somehow? E.g. introduce an enum and parse it once - with a failure if it is neither of the 3 allowed values.

File: sql/src/main/java/io/crate/expression/scalar/string/TrimFunctions.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import io.crate.expression.scalar.TripleScalar;
+import io.crate.metadata.FunctionIdent;
+import io.crate.metadata.FunctionInfo;
+import io.crate.types.DataTypes;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.HashSet;
+
+public final class TrimFunction {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/string/TrimFunctions.java]: Why is this here twice?

Comment (smakalias) [reply sql/src/main/java/io/crate/expression/scalar/string/TrimFunctions.java]: pushed this by mistake, I was trying something else -- need to stop using `git commit -a`
I'll remove

File: sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java (hunk 1 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java]: I'll be honest - checking this at runtime was what I had in mind :)
But this definitely sounds better 👍

File: sql-parser/src/main/antlr/SqlBase.g4 (hunk 2 above)
Comment (smakalias) [reply sql-parser/src/main/antlr/SqlBase.g4]: Dear me, I changed it in the docs and forgot to do so in the parser
good catch 👍

File: sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java (hunk 1 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java]: Would it also make sense to have an overload with 1 argument to also avoid the `mode` checks?

File: sql-parser/src/main/antlr/SqlBase.g4 (hunk 2 above)
Comment (smakalias) [reply sql-parser/src/main/antlr/SqlBase.g4]: @mfussenegger 
When I add the `TRIM` literal in the `nonReserved` List,  I get unknown function `trim('string')` when it is called with a single argument.

I noticed that other "special" functions are also not in the `nonReserved` List: `current_timestamp`, `case`, `extract`

When these are added, similar failures occur for these functions.

File: sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java (hunk 4 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java]: this is not working for me

File: sql-parser/src/main/antlr/SqlBase.g4 (hunk 2 above)
Comment (mfussenegger) [reply sql-parser/src/main/antlr/SqlBase.g4]: If we can't find a way to make it work with it being `nonReserved` we have to add a note that `trim` is now a reserved keyword to the breaking changes.

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 7]
```diff
@@ -1723,6 +1732,20 @@ private static String getUnquotedText(@Nullable ParserRuleContext context) {
         }
     }
 
+    private String getTrimMode(Token type) {
+        if (type == null) {
+            return "BOTH";
+        }
+        switch (type.getType()) {
+            case SqlBaseLexer.BOTH:
```
Comment (mfussenegger) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: Would it make sense to have the new enum in the sql-parser module and use it here already?

File: sql-parser/src/main/antlr/SqlBase.g4 (hunk 2 above)
Comment (smakalias) [reply sql-parser/src/main/antlr/SqlBase.g4]: fixed with latest fixup

File: sql-parser/src/main/java/io/crate/sql/tree/TrimModeEnum.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ * agreement.
+ */
+
+package io.crate.sql.tree;
+
+import java.util.Locale;
+
+public enum TrimModeEnum {
+
+    LEADING("LEADING"),
```
Comment (seut) [inline sql-parser/src/main/java/io/crate/sql/tree/TrimModeEnum.java]: if the value does not differ from the enum name, why not just use the name? no need for a value here..

File: sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+                } else {
+                    return new ThreeArgsTrimFunction(
+                        new FunctionInfo(new FunctionIdent(NAME, datatypes), datatypes.get(0))
+                    );
+                }
+            }
+        });
+    }
+
+    private abstract static class TrimBaseFunction extends Scalar<String, String> {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java]: I'd remove the base class - `getUniqueChars` isn't used in all sub-classes and could also be a static function. Having the class hierarchy for a simple `FunctionInfo` property is imho not worth it.

File: sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+            }
+            String charsToTrimArg = args[1].value();
+            if (charsToTrimArg == null) {
+                return target;
+            }
+            TrimMode mode = TrimMode.of(args[2].value());
+
+            HashSet<Character> charsToTrim = getUniqueChars(charsToTrimArg);
+
+            int start = 0;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java]: Might want to consider having something like `int start = mode.getStartIdx(charsToTrim)` and `int len = mode.getTrimmedLength(charsToTrim) `.

I think this would allow us to get rid of the two/three args distinction without adding any significant overhead.

Comment (kovrus) [reply sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java]: The second method needs to know about the start index and vice-versa, otherwise, pointers will cross. Also, both methods have to take the target string and probably we would need to have reloaded version of those methods for a single char. So, I am not sure whether it can be an improvement if I get the suggestion right.

Comment (mfussenegger) [reply sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java]: It would be mostly about removing the distinction of `if (mode.isTrailingOrBoth()) {` by using polymorphism.
